### DISCUSSION TITLE
readme enhancement: community supported builds (Windows / x64)

### DISCUSSION THREAD

**Discussion by franok** [async-profiler/async-profiler#502] [General]
Hi @apangin 

I would like to make a small suggestion to enhance the readme section about the [community supported Windows build](https://github.com/jvm-profiling-tools/async-profiler#community-supported-builds), but wanted to discuss it with you first.

In context of my master studies, I'm currently writing about Java profiling tools/technologies and came across async-profiler. During my evaluation I found that the section in the readme about Windows support is ~a little inprecise.~ EDIT: missing some information and thus could be misleading. By default, async-profiler will not work under Windows, although it is integrated in IntelliJ. Also [IntelliJ documentation](https://www.jetbrains.com/help/idea/cpu-profiler.html) does not explicitly mention that async-profiler integration works under Windows.
However, I figured out that it actually **is possible to use async-profiler under Windows** and get some useful results. All you need is WSL2:

I changed the run configurations in IntelliJ to execute my Java application on a linux build of an OpenJDK JVM which I installed on my WSL2 distro.

Based on my tryouts, I suggest to make the following addition to the readme:

---

### Community supported builds

 - **Windows** / x64 - <img src="https://upload.wikimedia.org/wikipedia/commons/9/9c/IntelliJ_IDEA_Icon.svg" width="16" height="16"/> [IntelliJ IDEA](https://www.jetbrains.com/idea/) 2021.2 and later.

Using the IntelliJ integration of async-profiler on a Windows machine requires execution on a linux platform build of any (OpenJDK) JVM for profiling a Java application. This can be achieved using a [WSL2](https://docs.microsoft.com/en-us/windows/wsl/) distro, since IntelliJ for Windows [supports WSL2 integration](https://www.jetbrains.com/help/idea/how-to-use-wsl-development-environment-in-product.html).

Running async-profiler from IntelliJ on a JVM for the Windows platform will not produce any flame graphs, since  it's not able to collect profiling data from perf counters.

---

 I know Windows support is not in the project scope, but guess it's probably worth mentioning this little detail, if people are interested in trying out the IntelliJ integration of async-profiler on a Windows machine (e.g. if they are bound to a Windows machine in their daily work - like I am for reasons...).

What do you think? 🙂 

If this is of any help, just let me know and I can create a PR to update the readme... or you can actually also directly copy the markdown snippet from this discussion if you prefer so.

Regards,
Frank

**Comment by gorrus** [comment]
Hi @franok , what do you mean by "By default, async-profiler will not work under Windows, although it is integrated in IntelliJ"? In Intellij IDEA 2021.2 it works on windows without wsl or any other stuff.

**Reply by franok** [reply]
Hi @gorrus, that's interesting. For me it didn't work using an OpenJDK 11 JVM for Windows in the IntelliJ run configs. I was able to start the application with async-profiler, but when I stopped it, it didn't produce any results:
![image](https://user-images.githubusercontent.com/17296805/141505698-f7a1d376-a173-4df1-b942-905ce251de07.png)

(using IntelliJ 2021.2.2 Ultimate)

So I figured out that switching to the JVM installed on my WSL2 made it work, showing flame graphs.
Now I wonder if I am doing anything wrong? 🤔 


However, I then still ask: why [IntelliJ does not mention](https://www.jetbrains.com/help/idea/cpu-profiler.html) that async-profiler also works with Windows ("by default"), but only on Linux and MacOS?:
> Async Profiler: a CPU and memory profiling tool for Linux and macOS.

**Reply by franok** [reply]
also: wouldn't async-profiler require access to _perf_ (perf_events) to work properly? afaik, there's no Windows port of perf, but it is inlcuded in the WSL2-Linux-Kernel: [tools/perf](https://github.com/microsoft/WSL2-Linux-Kernel/tree/linux-msft-wsl-5.10.y/tools/perf)

**Reply by gorrus** [reply]
This looks like a big, please file to the [IDEA bug tracker](https://youtrack.jetbrains.com/issues/IDEA), subsystem - java profiler cpu.
As for the [IDEA documentation](https://www.jetbrains.com/help/idea/cpu-profiler.html) - it will be fixed shortly, thanks!

**Reply by franok** [reply]
Ok, I created https://youtrack.jetbrains.com/issue/IDEA-282483

**Comment by franok** [comment]
Following the conversation with @gorrus this might be an IntelliJ bug in conjunction with missing documentation on IntelliJ side.

Probably, this gh-discussion can be closed.

**Reply by apangin** [reply]
Yes, IDEA's async-profiler is expected to work on Windows natively.
